tokio/src/sync/mpsc/chan.rs

```diff
@@ -199,6 +199,15 @@ where
     pub(crate) fn try_send(&mut self, value: T) -> Result<(), (T, TrySendError)> {
         self.inner.try_send(value, &mut self.permit)
     }
+
+    /// Returns `true` if the channel is closed.
+    ///
+    /// The channel is closed if `Rx::close` is called.
+    pub(crate) fn is_closed(&self) -> bool {
+        self.inner
+            .rx_fields
+            .with(|rx_fields_ptr| unsafe { (*rx_fields_ptr).rx_closed })
```

Comment:
I do not believe that this is safe.  There is no synchronization around `rx_fields`. This could be verified by adding a loom test that concurrently checks `is_closed` with calling `Receiver::close()`.

Instead, "closed" should be checked on `semaphore`.

Reply:
Thank you.

Following your suggestion I delved into the the `loom` tests and I indeed confirmed that `loom` complains with:

```
test sync::tests::loom_mpsc::is_closed_tx ... 
thread 'main' panicked at 'Causality violation: 
Concurrent read and write accesses.
```

The test is:

```rust
#[test]
fn is_closed_tx() {
    loom::model(|| {
        let (mut tx, mut rx) = mpsc::channel(16);

        let mut tx1 = tx.clone();
        thread::spawn(move || {
            assert!(!tx1.is_closed());
            tx1.try_send(()).unwrap();
        });

        let v = block_on(poll_fn(|cx| rx.poll_recv(cx)));
        assert!(v.is_some());
        rx.close();

        thread::spawn(move || {
            assert!(tx.is_closed());
            match tx.try_send(()) {
                Err(mpsc::error::TrySendError::Closed(_)) => {}
                _ => panic!("channel should be closed"),
            }
        });
    });
}
```

Is this the proper way to use a `loom::model`?

I then tried to protect the read of the `rx_fields` using the `Tx`'s permit without much success maybe because my (yet) limited familiarity with the internals of `tokio::sync::Semaphore` and `tokio::sync::SemaphorePermit`.

My questions: for a `Tx` to read the `rx_fields` is it the correct approach to use the `Tx`'s permit with `try_acquire`? 

Why the `Rx` instead does not have a permit? I understand that the `rx_fields` are strongly designed for exclusive read (perhaps the  idea of reading `rx_fields` from a `Tx` unfeasible)?

If so, how _"'closed' should be checked on semaphore"_ should be implemented?

Thank you in advance.

Reply:
When receiver closes the channel it also closes the corresponding semaphore. Senders could examine if semaphore is closed in thread-safe manner (instead of accessing `rx_closed` directly).